Netty: `Http2OrHttpHandler.handlerRemoved()` has thrown an exception

### Expected Behavior

N/A

### Actual Behaviour

Experiencing an error after upgrading to Micronaut `2.5.12`, along with Netty `4.1.66.Final` / TCNative `2.1.40.Final`. The exception seems innocuous in that it is logged, but does not cause any perceivable failure otherwise (our health checks pass, and regular traffic seems to pass just fine).

The exception:
```
io.netty.channel.ChannelPipelineException: io.micronaut.http.server.netty.NettyHttpServer$Http2OrHttpHandler.handlerRemoved() has thrown an exception.
        at io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0 (DefaultChannelPipeline.java:640)
        at io.netty.channel.DefaultChannelPipeline.remove (DefaultChannelPipeline.java:477)
        at io.netty.channel.DefaultChannelPipeline.remove (DefaultChannelPipeline.java:417)
        at io.netty.handler.ssl.ApplicationProtocolNegotiationHandler.removeSelfIfPresent (ApplicationProtocolNegotiationHandler.java:164)
        at io.netty.handler.ssl.ApplicationProtocolNegotiationHandler.userEventTriggered (ApplicationProtocolNegotiationHandler.java:143)
        at io.micronaut.http.server.netty.NettyHttpServer$Http2OrHttpHandler.userEventTriggered (NettyHttpServer.java:677)
        at io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered (AbstractChannelHandlerContext.java:346)
        at io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered (AbstractChannelHandlerContext.java:332)
        at io.netty.channel.AbstractChannelHandlerContext.fireUserEventTriggered (AbstractChannelHandlerContext.java:324)
        at io.netty.handler.ssl.SslHandler.setHandshakeSuccess (SslHandler.java:1837)
        at io.netty.handler.ssl.SslHandler.unwrap (SslHandler.java:1361)
        at io.netty.handler.ssl.SslHandler.decodeNonJdkCompatible (SslHandler.java:1248)
        at io.netty.handler.ssl.SslHandler.decode (SslHandler.java:1288)
        at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection (ByteToMessageDecoder.java:507)
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode (ByteToMessageDecoder.java:446)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead (ByteToMessageDecoder.java:276)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead (AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead (AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead (AbstractChannelHandlerContext.java:357)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead (DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead (AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead (AbstractChannelHandlerContext.java:365)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead (DefaultChannelPipeline.java:919)
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady (AbstractEpollStreamChannel.java:795)
        at io.netty.channel.epoll.EpollEventLoop.processReady (EpollEventLoop.java:480)
        at io.netty.channel.epoll.EpollEventLoop.run (EpollEventLoop.java:378)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run (SingleThreadEventExecutor.java:986)
        at io.netty.util.internal.ThreadExecutorMap$2.run (ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run (FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run (Thread.java:829)
Caused by: java.lang.IllegalStateException: recycled already
        at io.netty.util.Recycler$DefaultHandle.recycle (Recycler.java:242)
        at io.netty.util.internal.RecyclableArrayList.recycle (RecyclableArrayList.java:147)
        at io.netty.handler.ssl.ApplicationProtocolNegotiationHandler.handlerRemoved (ApplicationProtocolNegotiationHandler.java:95)
        at io.netty.channel.AbstractChannelHandlerContext.callHandlerRemoved (AbstractChannelHandlerContext.java:946)
        at io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0 (DefaultChannelPipeline.java:637)
```

Obviously, this exception seems to mention a lot of Netty stuff and no Micronaut stuff. I'm filing here because I figure the erroring callchain in question was setup by Micronaut, but please feel free to inform me otherwise if this is Netty related and should be filed somewhere else.

### Steps To Reproduce

1. Install Micronaut `2.5.12`
2. Install Netty `4.1.66.Final`
3. Install Netty TCNative `2.1.40.Final`
4. Setup SSL endpoint in Micronaut, ensuring it selects TCNative properly
5. Observe errors in log

### Environment Information

- GraalVM, regular JVM mode (HotSpot)
- Debian 10 / macOS 11 (occurs both in dev + production)
- JDK 11 `source`/`target`

### Example Application

_No response_

### Version

2.5.12

I have run into this same (?) problem and have a reproduction:

https://github.com/jonnii/http2error

Our environment is java 16 and this happens whenever the load balance hits the health endpoint - it goes away if you disable http2 (obviously) but it's not ideal.

One other observation we saw this both during the upgrade to 3.0 and also from 2.5.8 to 2.5.12 - the netty package upgrade between those two versions in the 2.x series was `4.1.64.Final` to `4.1.67.Final`.

In our case the no request state present error is preceded by a certificate_unknown error:

```
22:04:46.791 [default-nioEventLoopGroup-3-1] WARN  i.n.h.s.ApplicationProtocolNegotiationHandler - [id: 0x2494825a, L:/[0:0:0:0:0:0:0:1]:8443 ! R:/[0:0:0:0:0:0:0:1]:53481] TLS handshake failed:
javax.net.ssl.SSLHandshakeException: Received fatal alert: certificate_unknown
        at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:131)
        at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:117)
        at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:356)
        at java.base/sun.security.ssl.Alert$AlertConsumer.consume(Alert.java:293)
        at java.base/sun.security.ssl.TransportContext.dispatch(TransportContext.java:202)
        at java.base/sun.security.ssl.SSLTransport.decode(SSLTransport.java:171)
        at java.base/sun.security.ssl.SSLEngineImpl.decode(SSLEngineImpl.java:736)
        at java.base/sun.security.ssl.SSLEngineImpl.readRecord(SSLEngineImpl.java:691)
        at java.base/sun.security.ssl.SSLEngineImpl.unwrap(SSLEngineImpl.java:506)
        at java.base/sun.security.ssl.SSLEngineImpl.unwrap(SSLEngineImpl.java:482)
        at java.base/javax.net.ssl.SSLEngine.unwrap(SSLEngine.java:679)
        at io.netty.handler.ssl.JdkSslEngine.unwrap(JdkSslEngine.java:92)
        at io.netty.handler.ssl.JdkAlpnSslEngine.unwrap(JdkAlpnSslEngine.java:163)
        at io.netty.handler.ssl.SslHandler$SslEngineType$3.unwrap(SslHandler.java:297)
        at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1343)
        at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1236)
        at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1285)
        at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:507)
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:446)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:831)
22:04:46.863 [default-nioEventLoopGroup-3-2] ERROR i.m.h.s.netty.RoutingInBoundHandler - Micronaut Server Error - No request state present. Cause: io.micronaut.http.server.netty.NettyHttpServer$Http2OrHttpHandler.handlerRemoved() has thrown an exception.
io.netty.channel.ChannelPipelineException: io.micronaut.http.server.netty.NettyHttpServer$Http2OrHttpHandler.handlerRemoved() has thrown an exception.
        at io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0(DefaultChannelPipeline.java:640)
        at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:477)
        at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:417)
        at io.netty.handler.ssl.ApplicationProtocolNegotiationHandler.removeSelfIfPresent(ApplicationProtocolNegotiationHandler.java:173)
        at io.netty.handler.ssl.ApplicationProtocolNegotiationHandler.userEventTriggered(ApplicationProtocolNegotiationHandler.java:152)
        at io.micronaut.http.server.netty.NettyHttpServer$Http2OrHttpHandler.userEventTriggered(NettyHttpServer.java:725)
        at io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered(AbstractChannelHandlerContext.java:346)
        at io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered(AbstractChannelHandlerContext.java:332)
        at io.netty.channel.AbstractChannelHandlerContext.fireUserEventTriggered(AbstractChannelHandlerContext.java:324)
        at io.netty.handler.ssl.SslHandler.setHandshakeSuccess(SslHandler.java:1836)
        at io.netty.handler.ssl.SslHandler.wrapNonAppData(SslHandler.java:951)
        at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1409)
        at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1236)
        at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1285)
        at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:507)
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:446)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: java.lang.IllegalStateException: recycled already
        at io.netty.util.Recycler$DefaultHandle.recycle(Recycler.java:242)
        at io.netty.util.internal.RecyclableArrayList.recycle(RecyclableArrayList.java:147)
        at io.netty.handler.ssl.ApplicationProtocolNegotiationHandler.handlerRemoved(ApplicationProtocolNegotiationHandler.java:96)
        at io.netty.channel.AbstractChannelHandlerContext.callHandlerRemoved(AbstractChannelHandlerContext.java:946)
        at io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0(DefaultChannelPipeline.java:637)
        ... 32 common frames omitted
```

--- a/http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyHttpServer.java
+++ b/http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyHttpServer.java
@@ -663,43 +663,43 @@ public class NettyHttpServer implements NettyEmbeddedServer {
         if (childHandler != null) {
             final ServerSslBuilder serverSslBuilder = nettyEmbeddedServices.getServerSslBuilder();
             childHandler.sslContext = serverSslBuilder != null ? serverSslBuilder.build().orElse(null) : null;
-            childHandler.http2OrHttpHandler = new Http2OrHttpHandler(childHandler.sslContext != null, serverConfiguration.getFallbackProtocol());
             childHandler.loggingHandler = serverConfiguration.getLogLevel().isPresent() ?
                     new LoggingHandler(NettyHttpServer.class, serverConfiguration.getLogLevel().get()) : null;
+            childHandler.initHttpCoders();
         }
     }
 
     /**
      * Negotiates with the browser if HTTP2 or HTTP is going to be used. Once decided, the Netty
      * pipeline is setup with the correct handlers for the selected protocol.
+     *
+     * @implNote Unfortunately, this handler cannot be {@link io.netty.channel.ChannelHandler.Sharable shared} because
+     * {@link ApplicationProtocolNegotiationHandler} does not support it.
      */
-    @ChannelHandler.Sharable
     private final class Http2OrHttpHandler extends ApplicationProtocolNegotiationHandler {
         private final boolean useSsl;
-        // all are Sharable
-        final HttpAccessLogHandler accessLogHandler =
-                serverConfiguration.getAccessLogger() != null && serverConfiguration.getAccessLogger()
-                .isEnabled() ?
-                new HttpAccessLogHandler(serverConfiguration.getAccessLogger().getLoggerName(),
-                                         serverConfiguration.getAccessLogger().getLogFormat()) : null;
-        final HttpRequestDecoder requestDecoder = new HttpRequestDecoder(NettyHttpServer.this,
-                                                                         environment,
-                                                                         serverConfiguration,
-                                                                         httpRequestReceivedEventPublisher);
-        final HttpResponseEncoder responseDecoder = new HttpResponseEncoder(
-                nettyEmbeddedServices.getMediaTypeCodecRegistry(),
-                serverConfiguration
-        );
+        private final HttpAccessLogHandler accessLogHandler;
+        private final HttpRequestDecoder requestDecoder;
+        private final HttpResponseEncoder responseEncoder;
 
         /**
          * Default constructor.
          *
          * @param useSsl           true when using ssl
          * @param fallbackProtocol The fallback protocol
+         * @param accessLogHandler Handler for writing access logs
+         * @param requestDecoder   Request decoder
+         * @param responseEncoder  Response encoder
          */
-        Http2OrHttpHandler(boolean useSsl, String fallbackProtocol) {
+        Http2OrHttpHandler(boolean useSsl, String fallbackProtocol,
+                           @Nullable HttpAccessLogHandler accessLogHandler,
+                           HttpRequestDecoder requestDecoder,
+                           HttpResponseEncoder responseEncoder) {
             super(fallbackProtocol);
             this.useSsl = useSsl;
+            this.accessLogHandler = accessLogHandler;
+            this.requestDecoder = requestDecoder;
+            this.responseEncoder = responseEncoder;
         }
 
         @Override
@@ -797,7 +797,7 @@ public class NettyHttpServer implements NettyEmbeddedServer {
             if (useSsl) {
                 handlers.put("request-certificate-handler", requestCertificateHandler);
             }
-            handlers.put(HttpResponseEncoder.ID, responseDecoder);
+            handlers.put(HttpResponseEncoder.ID, responseEncoder);
             handlers.put(NettyServerWebSocketUpgradeHandler.ID, new NettyServerWebSocketUpgradeHandler(
                     nettyEmbeddedServices,
                     getWebSocketSessionRepository()
@@ -812,9 +812,31 @@ public class NettyHttpServer implements NettyEmbeddedServer {
      */
     private class NettyHttpServerInitializer extends ChannelInitializer<SocketChannel> {
         SslContext sslContext = nettyEmbeddedServices.getServerSslBuilder() != null ? nettyEmbeddedServices.getServerSslBuilder().build().orElse(null) : null;
-        Http2OrHttpHandler http2OrHttpHandler = new Http2OrHttpHandler(sslContext != null, serverConfiguration.getFallbackProtocol());
         LoggingHandler loggingHandler =
                 serverConfiguration.getLogLevel().isPresent() ? new LoggingHandler(NettyHttpServer.class, serverConfiguration.getLogLevel().get()) : null;
+
+        private HttpAccessLogHandler accessLogHandler;
+        private HttpRequestDecoder requestDecoder;
+        private HttpResponseEncoder responseEncoder;
+
+        {
+            initHttpCoders();
+        }
+
+        void initHttpCoders() {
+            this.accessLogHandler = serverConfiguration.getAccessLogger() != null && serverConfiguration.getAccessLogger()
+                    .isEnabled() ?
+                    new HttpAccessLogHandler(serverConfiguration.getAccessLogger().getLoggerName(),
+                            serverConfiguration.getAccessLogger().getLogFormat()) : null;
+            this.requestDecoder = new HttpRequestDecoder(NettyHttpServer.this,
+                    environment,
+                    serverConfiguration,
+                    httpRequestReceivedEventPublisher);
+            this.responseEncoder = new HttpResponseEncoder(
+                    nettyEmbeddedServices.getMediaTypeCodecRegistry(),
+                    serverConfiguration
+            );
+        }
 
         @Override
         protected void initChannel(SocketChannel ch) {
@@ -829,6 +851,8 @@ public class NettyHttpServer implements NettyEmbeddedServer {
             if (loggingHandler != null) {
                 pipeline.addLast(loggingHandler);
             }
+
+            Http2OrHttpHandler http2OrHttpHandler = new Http2OrHttpHandler(sslContext != null, serverConfiguration.getFallbackProtocol(), accessLogHandler, requestDecoder, responseEncoder);
 
             if (serverConfiguration.getHttpVersion() != io.micronaut.http.HttpVersion.HTTP_2_0) {
                 http2OrHttpHandler.configurePipeline(ApplicationProtocolNames.HTTP_1_1, pipeline);

